### PR TITLE
refactor(discovery-client): migrate Discovery Client source to TypeScript

### PR DESCRIPTION
**Pull request by mcous**
## overview

This PR takes the Discovery Client source code and migrates it to TypeScript!

~Currently blocked by #7274, but has been used as a development branch to figure out the proper TS configuration(s) in that PR.~ Blocking PR merged, this PR rebased and ready for review

## changelog

- Migrate DC to TypeScript
- Update build to include TS typechecking now that we have source files to check

### notes on the TS build system

I went with TypeScript's [Project References](https://www.typescriptlang.org/docs/handbook/project-references.html) build system for this based on my [previous TS-in-a-monorepo experience](https://github.com/tracespace/tracespace/tree/v5). The setup is a little bit more complex than a system where you run `tsc --noEmit "**/*.ts"`, but it brings the following benefits:

- Faster typechecks due to incremental builds
- Enables npm publishing capability that we lost due to how Flow expected things to be configured

Key details:

- In addition to checking types, TS will emit declaration files (`*.d.ts`), which serve two purposes:
    1. Speed up future typechecks (in addition to `*.tsbuildinfo` cache files)
    2. Serve as the project's type information when published to npm
- We configure `package.json` to point `main` at compiled code and `types` at declaration files
    - Upside: since `main` points at compiled code, the package is now able to be published to npm 
    - Downside: since `main` points at compiled code, dependency projects' tests and bundle builds will try to pull compiled code
    - This means, by default, we would need to bundle DC with webpack before `app` or `app-shell` tests would be able to work
- Needing to bundle before running tests and dependent bundles is a _huge_ pain in my experience
    - Babel to the rescue! We use [babel-plugin-module-resolver](https://github.com/tleunen/babel-plugin-module-resolver) to rewrite `@opentrons/discovery-client` import statements to `@opentrons/discovery-client/src/index.ts`
    - Now, tests and bundles from other projects (e.g. `@opentrons/app-shell`) that pull in DC stuff will pull in the source code directly, leading to tests that run without builds and more efficient bundles
    - As configured, this relies on dependent projects always importing from the module name, which is the right thing to do, anyway
- Another minor downside: projects configured this way can only import files from their `projectRoot`, which is usually `src`
    - Before this, DC imported its version from `package.json` _outside of `src`_, which no longer works
    - My preferred workaround here is to use webpack to bundle in whatever we need from `package.json` as globals, e.g. `_PKG_VERSION_`
    - There are other workarounds involving multiple `tsconfig.json` files, but the webpack thing is already something we do, so it feels cleaner

## review requests

~This PR is still in a draft state because it makes a few changes to the source code that should be reverted in favor of a lint ignore comment or similar.~ Please look through and try to identify anything that should be an ignore comment rather than a code change.

Please also call out anything that looks like it should go in our Flow > TS conversion guide!

## risk assessment

Medium. The Discovery Client is heavily unit tested, so we've have an amount of regression safety (acknowledging that some of these tests mock out 3rd-party dependencies, reducing their value and safety).

Smoke tests are prudent, but the nature of the project is such that if the smoke tests look good, we're probably ok.

### CODE REVIEW

**Review by IanLondon — APPROVED** [review]
code/config changes seems cool to me. Did not test bc running `yarn` is taking forever lately on this machine, LMK if there's anything you'd like to me specifically test

File: discovery-client/src/cli.ts [hunk 1]
```diff
@@ -47,10 +49,9 @@ const createLogger = (argv: Argv): Logger => {
   }
 }
 
-const debugLogArgvMiddleware = (argv: Argv) => {
+const debugLogArgvMiddleware: MiddlewareFunction<Argv> = (argv): void => {
   const log = createLogger(argv)
   log.debug(`Calling ${argv.$0} with argv:`, argv)
-  return argv
```
Comment (mcous) [inline discovery-client/src/cli.ts:53]: Throw a TODO and ignore on here instead of removing this return

File: discovery-client/src/cli.ts [hunk 2]
```diff
@@ -101,12 +102,12 @@ const browse = (argv: Argv) => {
...
 }
 
-const find = (argv: FindArgv) => {
+const find = (argv: FindArgv): void => {
   const { name, timeout, ipFilter } = argv
   const log = createLogger(argv)
   const client = createClient(argv, robots => {
     robots
-      .filter(robot => !name || robot.name === name)
+      .filter(robot => name == null || robot.name === name)
```
Comment (mcous) [inline discovery-client/src/cli.ts]: Ignore rather than code change

File: discovery-client/src/cli.ts [hunk 3]
```diff
@@ -120,7 +121,9 @@ const find = (argv: FindArgv) => {
   })
 
   log.warn(
-    `Finding ${argv.name ? `robot "${argv.name}"` : 'first available robot'}`
+    `Finding ${
+      argv.name != null ? `robot "${argv.name}"` : 'first available robot'
```
Comment (mcous) [inline discovery-client/src/cli.ts]: Ignore rather than code change

File: discovery-client/src/cli.ts [hunk 4]
```diff
@@ -158,7 +161,7 @@ Yargs.options({
   logLevel: {
     describe: 'Log level',
     alias: 'l',
-    choices: [...LOG_LVLS, 'off'],
+    choices: LOG_LVLS,
```
Comment (mcous) [inline discovery-client/src/cli.ts]: Revert this change

File: discovery-client/src/discovery-client.ts [hunk 5]
```diff
@@ -46,7 +47,7 @@ export function createDiscoveryClient(
     healthPoller.start({ list: prevAddrs, interval: healthPollInterval })
     mdnsBrowser.start()
 
-    if (!unsubscribe) {
+    if (unsubscribe === null) {
```
Comment (mcous) [inline discovery-client/src/discovery-client.ts]: Ignore rather than code change

File: discovery-client/src/discovery-client.ts [hunk 6]
```diff
@@ -63,7 +64,7 @@ export function createDiscoveryClient(
   const stop = (): void => {
     mdnsBrowser.stop()
     healthPoller.stop()
-    if (unsubscribe) {
+    if (typeof unsubscribe === 'function') {
```
Comment (mcous) [inline discovery-client/src/discovery-client.ts]: Ignore rather than code change

File: discovery-client/src/health-poller.ts [hunk 7]
```diff
@@ -80,7 +83,7 @@ export function createHealthPoller(options: HealthPollerOptions): HealthPoller {
 
     // if xor (symmetric difference) returns values, then elements exist in
     // one list and not the other and need to be added to and/or removed from the queue
-    if (nextList && xorBy(pollQueue, nextList, 'ip').length > 0) {
+    if (nextList != null && xorBy(pollQueue, nextList, 'ip').length > 0) {
```
Comment (mcous) [inline discovery-client/src/health-poller.ts]: Ignore rather than code change

File: discovery-client/src/health-poller.ts [hunk 8]
```diff
@@ -92,14 +95,16 @@ export function createHealthPoller(options: HealthPollerOptions): HealthPoller {
     }
 
     if (needsNewInterval && pollQueue.length > 0 && interval > 0) {
-      const handlePoll = () => {
+      const handlePoll = (): void => {
+        const head = pollQueue.shift()
```
Comment (mcous) [inline discovery-client/src/health-poller.ts]: Ignore rather than code change

File: discovery-client/src/health-poller.ts [hunk 9]
```diff
@@ -111,19 +116,22 @@ export function createHealthPoller(options: HealthPollerOptions): HealthPoller {
...
   }
 
-  const stop = () => {
+  function stop(): void {
     log('debug', 'stopping health poller')
     lastCompletedPollTimeByIp = {}
-    clearInterval(pollIntervalId)
-    pollIntervalId = null
+
+    if (pollIntervalId !== null) {
```
Comment (mcous) [inline discovery-client/src/health-poller.ts]: Ignores rather than code changes

File: discovery-client/src/mdns-browser.ts [hunk 10]
```diff
@@ -1,38 +1,29 @@
...
-  BrowserService,
-  NetworkConnection,
-} from 'mdns-js'
 
+import type { Browser as BaseBrowser, BrowserService } from 'mdns-js'
 import type { MdnsBrowser, MdnsBrowserOptions, LogLevel } from './types'
 
 monkeyPatchThrowers()
 
-// TODO(mc, 2020-07-14): remove the function in favor of createMdnsBrowser
```
Comment (mcous) [inline discovery-client/src/mdns-browser.ts:18]: I think these removals are safe because these functions were completely unused. Basically, should've been removed months ago as part of the big DC refactor but they were missed

File: discovery-client/src/mdns-browser.ts [hunk 11]
```diff
@@ -64,8 +55,8 @@ export function createMdnsBrowser(options: MdnsBrowserOptions): MdnsBrowser {
     browser = baseBrowser
   }
 
-  const stop = () => {
-    if (browser) {
+  function stop(): void {
+    if (browser !== null) {
```
Comment (mcous) [inline discovery-client/src/mdns-browser.ts]: Ignore instead of code change

File: discovery-client/src/store/reducer.ts [hunk 12]
```diff
@@ -41,20 +51,24 @@ const makeHostState = (ip, port, robotName) => ({
...
 })
 
-const getHealthStatus = (responseData, error) => {
-  if (responseData) return HEALTH_STATUS_OK
-  if (error && error.status >= 400) return HEALTH_STATUS_NOT_OK
+const getHealthStatus = (
+  responseData: HealthResponse | ServerHealthResponse | null,
+  error: HealthErrorResponse | null
+): HealthStatus => {
+  if (responseData !== null) return HEALTH_STATUS_OK
```
Comment (mcous) [inline discovery-client/src/store/reducer.ts]: Ignore rather than code change for these two lines

File: discovery-client/src/store/reducer.ts [hunk 13]
```diff
@@ -41,20 +51,24 @@ const makeHostState = (ip, port, robotName) => ({
...
 export const robotsByNameReducer = (
+  // eslint-disable-next-line @typescript-eslint/default-param-last
   state: RobotsByNameMap = INITIAL_STATE.robotsByName,
   action: Action
 ): RobotsByNameMap => {
   switch (action.type) {
     case Actions.INITIALIZE_STATE: {
       const { initialRobots } = action.payload
-      if (!initialRobots) return state
+      if (initialRobots == null) return state
```
Comment (mcous) [inline discovery-client/src/store/reducer.ts]: Ignore rather than code change

File: discovery-client/src/store/reducer.ts [hunk 14]
```diff
@@ -81,7 +95,7 @@ export const robotsByNameReducer = (
     case Actions.HEALTH_POLLED: {
       const { health, serverHealth } = action.payload
       const name = serverHealth?.name ?? health?.name ?? null
-      if (!name) return state
+      if (name === null) return state
```
Comment (mcous) [inline discovery-client/src/store/reducer.ts]: Ignore rather than code change

File: discovery-client/src/store/reducer.ts [hunk 15]
```diff
@@ -99,13 +113,14 @@ export const robotsByNameReducer = (
...
 export const hostsByIpReducer = (
+  // eslint-disable-next-line @typescript-eslint/default-param-last
   state: HostsByIpMap = INITIAL_STATE.hostsByIp,
   action: Action
 ): HostsByIpMap => {
   switch (action.type) {
     case Actions.INITIALIZE_STATE: {
       const { initialRobots } = action.payload
-      if (!initialRobots) return state
+      if (initialRobots == null) return state
```
Comment (mcous) [inline discovery-client/src/store/reducer.ts]: Ignore rather than code change

File: discovery-client/src/store/reducer.ts [hunk 16]
```diff
@@ -162,7 +177,7 @@ export const hostsByIpReducer = (
         serverHealthError
       )
       const seen =
-        host?.seen === true ||
+        host?.seen ||
```
Comment (mcous) [inline discovery-client/src/store/reducer.ts]: Ignore rather than code change

File: discovery-client/src/store/reducer.ts [hunk 17]
```diff
@@ -195,7 +210,7 @@ export const hostsByIpReducer = (
             return (
               targetIp !== ip &&
               targetRobotName === robotName &&
-              targetSeen === false &&
+              !targetSeen &&
```
Comment (mcous) [inline discovery-client/src/store/reducer.ts]: Ignore rather than code change

**Review by shlokamin — APPROVED** [review]
Seems reasonable to me. Approving pending CI. I also agree that we should minimize code changes in favor of lint ignores.

File: app/src/redux/discovery/constants.js [hunk 18]
```diff
@@ -8,7 +8,7 @@ export {
   RE_HOSTNAME_IPV4_LL,
   RE_HOSTNAME_LOCALHOST,
   RE_HOSTNAME_LOOPBACK,
-} from '@opentrons/discovery-client/src/constants'
```
Comment (mcous) [inline app/src/redux/discovery/constants.js:11]: This change must be reverted to build successfully

File: discovery-client/src/store/reducer.ts [hunk 19]
```diff
@@ -212,9 +234,10 @@ export const hostsByIpReducer = (
 }
 
 export const manualAddressesReducer = (
-  state: $ReadOnlyArray<Address> = INITIAL_STATE.manualAddresses,
+  // eslint-disable-next-line @typescript-eslint/default-param-last
+  state: Address[] = INITIAL_STATE.manualAddresses,
   action: Action
-): $ReadOnlyArray<Address> => {
+): Address[] => {
```
Comment (b-cooper) [inline discovery-client/src/store/reducer.ts:240]: can/should this be a `ReadonlyArray`?

**Review by b-cooper — APPROVED** [review]
Looks great 🍖 

Also, I'll add a little blurb about migrating `$ReadOnlyArray<ElemType>` to `ElemType[]` or `ReadonlyArray<ElemType>` differentially.

File: discovery-client/src/store/reducer.ts (hunk 19 above)
Comment (mcous) [reply discovery-client/src/store/reducer.ts:240]: Yeah, I think it could! I think it also depends on how well (or not) it plays with the `combineReducers` typing
